File: kangaroo/kangaroo.clj
(defn kangaroo [x-loc, x-jump, y-loc, y-jump]
  (let [[behind-jump, ahead-jump]
        (if (< x-loc y-loc) [x-jump y-jump] [y-jump x-jump])
        is-no? (>= ahead-jump behind-jump)
        [behind, ahead] (sort [x-loc y-loc])
        get-is-yes? (fn get-is-yes? [curr-behind curr-ahead]
                      (if (<= curr-ahead curr-behind)
                        (= curr-ahead curr-behind)
                        (get-is-yes?
                         (+ curr-behind behind-jump)
                         (+ curr-ahead ahead-jump))))]
    (cond
      is-no? "NO"
      (get-is-yes? behind ahead) "YES"
      :else "NO")))

File: kangaroo/kangaroo.js
// https://www.hackerrank.com/challenges/kangaroo/problem

const kangaroo = (xLoc, xJump, yLoc, yJump) => {
  const [behindJump, aheadJump] = xLoc > yLoc ? [yJump, xJump] : [xJump, yJump];

  if (aheadJump >= behindJump) return 'NO';

  const [behind, ahead] = [xLoc, yLoc].sort();

  const isYes = (function getIsYes(currBehind, currAhead) {
    return currAhead <= currBehind
      ? currAhead === currBehind
      : getIsYes(currBehind + behindJump, currAhead + aheadJump);
  })(behind, ahead);

  return isYes ? 'YES' : 'NO';
};
